File: tools/sumaho_gate.mjs
#!/usr/bin/env node
/**
 * ★1043番【スマホの門】たまごさんのiPhoneのSafariと同じ条件で、箱のページを開いて
 * 実際にボタンを押し、届いたかどうかを画面から読み取る門。
 *
 * なぜ要るか（たまごさん原文 2026-09-24）:
 *   「こういうのを俺で試さないでよ。俺が依頼したものが確実になってるものだけ見せてよ」
 *   「俺が入れたもの全然入ってないよ」
 *   → 1038〜1042番の実測は全部 curl / python で叩いていた。**経路が違う。**
 *     ブラウザは独自ヘッダが付くと本番の前に preflight(OPTIONS) を投げる。curlは投げない。
 *     だから「機械からは通る／たまごさんからは1件も届かない」が起きた。
 *     この門は**ブラウザから押す。**preflightも走る。CORSで止まればここで止まる。
 *
 * 安全側（tools/_825_screenshot_375.mjs と同じ型をそのまま継承する）:
 *   ・headless Chrome ／ 使い捨てプロファイル ／ 終わったら必ず殺す
 *   ・★たまごさんの普段のブラウザには一切触らない。タブを1枚も増やさない
 *   ・投げるURLには ?kikai=1 を付ける＝台帳に「機械の試し投げ」の印が入る
 *     （たまごさんの一覧には1件も混ざらない）
 *
 * 使い方:
 *   node tools/sumaho_gate.mjs <箱のURL> <出力JSONパス> [PNGパス]
 */
import { spawn } from "node:child_process";
import { mkdtempSync, mkdirSync, existsSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, resolve } from "node:path";

const BOX = process.argv[2];
const OUT = resolve(process.argv[3] || "/tmp/sumaho_gate.json");
const PNG = process.argv[4] ? resolve(process.argv[4]) : null;
if (!BOX) { console.error("使い方: node tools/sumaho_gate.mjs <箱のURL> <出力JSON> [PNG]"); process.exit(1); }

const CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome";
// ★iPhoneのSafariのUA（iOS 17）。375x812＝iPhone SE/13 miniの実寸
const UA = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_5 like Mac OS X) AppleWebKit/605.1.15 "
         + "(KHTML, like Gecko) Version/17.5 Mobile/15E148 Safari/604.1";
const W = 375, H = 812;
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function findPage(port) {
  for (let i = 0; i < 160; i++) {
    try {
      const list = await (await fetch(`http://127.0.0.1:${port}/json/list`)).json();
      const page = list.find((t) => t.type === "page");
      if (page?.webSocketDebuggerUrl) return page;
    } catch { /* 起動待ち */ }
    await sleep(250);
  }
  throw new Error("Chromeのデバッグ口が開きませんでした");
}

function connect(wsUrl, onEvent) {
  const ws = new WebSocket(wsUrl);
  let id = 0;
  const pending = new Map();
  ws.addEventListener("message", (ev) => {
    const msg = JSON.parse(ev.data);
    if (msg.id && pending.has(msg.id)) { const cb = pending.get(msg.id); pending.delete(msg.id); cb(msg); }
    else if (msg.method && onEvent) onEvent(msg);
  });
  const ready = new Promise((res, rej) => {
    ws.addEventListener("open", res);
    ws.addEventListener("error", (e) => rej(new Error(`CDP接続失敗: ${e?.message ?? e}`)));
  });
  const send = (method, params = {}) => new Promise((res, rej) => {
    const myId = ++id;
    pending.set(myId, (m) => (m.error ? rej(new Error(JSON.stringify(m.error))) : res(m.result)));
    ws.send(JSON.stringify({ id: myId, method, params }));
  });
  return { ready, send, close: () => ws.close() };
}

async function evalIn(send, expr) {
  const r = await send("Runtime.evaluate", { expression: expr, returnByValue: true, awaitPromise: true });
  return r?.result?.value;
}

async function main() {
  const res = { at: new Date().toISOString(), box: BOX, ua: "iPhone Safari 17.5", size: `${W}x${H}`,
                console: [], netFail: [], steps: [], ok: false };
  const port = 9600 + Math.floor(Math.random() * 300);
  const profile = mkdtempSync(resolve(tmpdir(), "sumaho-gate-"));
  const chrome = spawn(CHROME, [
    "--headless=new", `--remote-debugging-port=${port}`, `--user-data-dir=${profile}`,
    "--no-first-run", "--mute-audio", `--window-size=${W},${H}`,
    "--disable-features=Translate,MediaRouter", "about:blank",
  ], { stdio: "ignore" });

  try {
    const target = await findPage(port);
    const { ready, send, close } = connect(target.webSocketDebuggerUrl, (msg) => {
      // ★スマホではconsoleが見えない。ここで全部拾って証拠に残す
      if (msg.method === "Runtime.consoleAPICalled") {
        res.console.push((msg.params.type || "log") + ": " +
          (msg.params.args || []).map((a) => a.value ?? a.description ?? "").join(" ").slice(0, 300));
      }
      if (msg.method === "Log.entryAdded") {
        const e = msg.params.entry || {};
        res.console.push(`${e.level}(${e.source}): ${String(e.text || "").slice(0, 300)}`);
        // ★CORSで止められた場合はここに出る（これが1043番で探していたもの）
        if (/cors|Access-Control|preflight/i.test(e.text || "")) res.corsBlocked = String(e.text).slice(0, 400);
      }
      if (msg.method === "Network.loadingFailed") {
        res.netFail.push(`${msg.params.type} ${msg.params.errorText}${msg.params.corsErrorStatus
          ? " / CORS:" + JSON.stringify(msg.params.corsErrorStatus) : ""}`);
        if (msg.params.corsErrorStatus) res.corsBlocked = JSON.stringify(msg.params.corsErrorStatus);
      }
    });
    await ready;
    await send("Page.enable"); await send("Runtime.enable");
    await send("Log.enable"); await send("Network.enable");
    await send("Emulation.setDeviceMetricsOverride",
      { width: W, height: H, deviceScaleFactor: 3, mobile: true });
    await send("Emulation.setUserAgentOverride", { userAgent: UA, platform: "iPhone" });
    await send("Emulation.setTouchEmulationEnabled", { enabled: true, maxTouchPoints: 5 });

    const url = BOX + (BOX.includes("?") ? "&" : "?") + "kikai=1&t=" + Date.now();
    await send("Page.navigate", { url });
    for (let i = 0; i < 60 && (await evalIn(send, "document.readyState")) !== "complete"; i++) await sleep(300);
    await sleep(2500); // 棚一覧(tana_ichiran.json)のfetch待ち
    res.steps.push("箱を開いた（readyState complete）");

    const has = await evalIn(send, `JSON.stringify({
      send: !!document.getElementById("send"), u: !!document.getElementById("u"),
      m: !!document.getElementById("m"), nomu: !!document.getElementById("nomu"),
      tana: (document.getElementById("tanaNote")||{}).textContent || "" })`);
    res.parts = JSON.parse(has || "{}");
    if (!res.parts.send) throw new Error("［投 げ る］ボタンが画面に無い");
    res.steps.push("ボタンと欄がある: " + has);

    // ★4パターンを、たまごさんと同じ道（ブラウザ→中継所）で順に押す
    const stamp = Date.now().toString(36);
    const patterns = [
      { name: "①URLだけ（棚なし・ひとことなし）", u: "https://www.youtube.com/watch?v=J---aiyznGQ", m: "", tana: false },
      { name: "②URL＋棚", u: "https://www.youtube.com/watch?v=lDK9QqIzhwk", m: "", tana: true },
      { name: "③ひとことだけ", u: "", m: "1043番 スマホの門 ひとことだけ " + stamp, tana: false },
      { name: "④日本語の題名のYouTube", u: "https://www.youtube.com/watch?v=WSeNSzJ2-Jw", m: "1043番 日本語題名の実測 " + stamp, tana: false },
    ];
    res.oshita = [];
    for (const pt of patterns) {
      await evalIn(send, `(function(){
        document.getElementById("u").value = ${JSON.stringify(pt.u)};
        document.getElementById("m").value = ${JSON.stringify(pt.m)};
        document.getElementById("msg").textContent = "";
        ${pt.tana ? `var b=document.querySelector("#tana button"); if(b) b.click();` : ``}
        return 1; })()`);
      await sleep(250);
      // ★実際にボタンを押す（クリックのイベントをそのまま起こす）
      await evalIn(send, `document.getElementById("send").click(), 1`);
      let msg = "", cls = "";
      for (let i = 0; i < 50; i++) {
        await sleep(400);
        msg = await evalIn(send, `(document.getElementById("msg")||{}).textContent || ""`);
        cls = await evalIn(send, `(document.getElementById("msg")||{}).className || ""`);
        if (msg && !/送っています/.test(msg)) break;
      }
      const nokori = await evalIn(send, `(function(){try{return JSON.parse(localStorage.getItem("nagekomi.pending")||"[]").length}catch(e){return -1}})()`);
      res.oshita.push({ pattern: pt.name, gamen: String(msg).slice(0, 300), ok: /ok/.test(cls),
                        tanmatsuNokori: nokori });
      res.steps.push(`${pt.name} → 画面:「${String(msg).slice(0, 120)}」`);
      await sleep(600);
    }
    res.diag = await evalIn(send, `(document.getElementById("diag")||{}).textContent || ""`);
    res.tsuutta = res.oshita.filter((x) => x.ok).length;
    res.ok = res.tsuutta === patterns.length;

    if (PNG) {
      mkdirSync(dirname(PNG), { recursive: true });
      const shot = await send("Page.captureScreenshot", { format: "png", captureBeyondViewport: true });
      writeFileSync(PNG, Buffer.from(shot.data, "base64"));
      res.png = PNG;
    }
    close();
  } catch (e) {
    res.error = String(e && e.message ? e.message : e);
  } finally {
    try { chrome.kill("SIGKILL"); } catch {}
  }
  if (!existsSync(dirname(OUT))) mkdirSync(dirname(OUT), { recursive: true });
  writeFileSync(OUT, JSON.stringify(res, null, 1));
  console.log(JSON.stringify(res, null, 1));
  process.exit(res.ok ? 0 : 1);
}
main();

File: tools/sumaho_gate.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""★1043番【スマホの門】たまごさんのiPhoneのSafariと同じ経路で、箱のボタンを実際に押す門。

たまごさん（2026-09-24・原文）:
  「こういうのを俺で試さないでよ。俺が依頼したものが確実になってるものだけ見せてよ」
  「Macからの投げが通っても意味がない。たまごさんが使う経路で通す」

やること（工場側＝Macで走る。サンドボックスからは中継所へ出られないため）:
  1. tools/sumaho_gate.mjs を呼ぶ（headless Chrome・iPhoneのUA・375x812・タッチあり）
  2. 本番の箱ページを開いて、4パターンを**ボタンを押して**投げる
  3. 画面に出た文字（届いた／届かなかった＋理由）と、consoleのエラー、CORSの遮断を全部拾う
  4. 台帳（status/nagekomi.jsonl）が本当に増えたかを、投げる前後の行数で突き合わせる
     ★画面が「届いた」と言っても台帳が増えていなければ**通っていない**と判定する

★投げるURLには ?kikai=1 が付く＝台帳に「機械の試し投げ」の印が入る。
  たまごさんの一覧には1件も混ざらない（tools/nagekomi_list.py が既定で外す）。

使い方:
  python3 tools/sumaho_gate.py
  python3 tools/sumaho_gate.py --box <箱のURL>
"""
from __future__ import annotations

import argparse
import io
import json
import os
import subprocess
import sys
import time
from datetime import datetime, timedelta, timezone

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(HERE)
JST = timezone(timedelta(hours=9))
STATUS = os.path.join(REPO, "status")
LEDGER = os.path.join(STATUS, "nagekomi.jsonl")
OUT_JSON = os.path.join(STATUS, "sumaho_gate.json")
PNG = os.path.join(STATUS, "sumaho_gate_375.png")
BOX = ("https://tamago2022.github.io/tamago-shinchoku/share/"
       "nagekomi-c5fd9d5791b32e88.html")
LIMIT_SEC = 300


def now():
    return datetime.now(JST)


def ledger_ids():
    ids = set()
    try:
        for line in io.open(LEDGER, encoding="utf-8"):
            line = line.strip()
            if not line:
                continue
            try:
                ids.add(json.loads(line).get("id"))
            except Exception:
                pass
    except OSError:
        pass
    return ids


def run(box=None):
    box = box or BOX
    res = {"ranAt": now().strftime("%Y-%m-%d %H:%M"), "box": box,
           "diag": [], "red": [], "totalYen": 0.0}
    mjs = os.path.join(HERE, "sumaho_gate.mjs")
    if not os.path.exists(mjs):
        res["red"].append("tools/sumaho_gate.mjs が無い")
        res["ok"] = False
        return res

    before = ledger_ids()
    res["diag"].append("投げる前の台帳 %d行" % len(before))
    t0 = time.time()
    try:
        r = subprocess.run(["node", mjs, box, OUT_JSON, PNG],
                           capture_output=True, text=True, timeout=LIMIT_SEC)
    except subprocess.TimeoutExpired:
        res["red"].append("スマホの門が %d秒で終わらなかった" % LIMIT_SEC)
        res["ok"] = False
        return res
    res["seconds"] = round(time.time() - t0, 1)
    try:
        gate = json.load(io.open(OUT_JSON, encoding="utf-8"))
    except Exception:
        gate = {"error": (r.stderr or r.stdout or "")[-600:]}
    res["gamen"] = gate.get("oshita") or []
    res["console"] = (gate.get("console") or [])[-12:]
    res["netFail"] = (gate.get("netFail") or [])[-8:]
    res["pageDiag"] = str(gate.get("diag") or "")[:800]
    if gate.get("corsBlocked"):
        # ★これが1043番で探していたもの。ブラウザが本番のPOSTを送らずに捨てた証拠
        res["red"].append("★ブラウザが止めた（CORS）：%s" % str(gate["corsBlocked"])[:200])
    if gate.get("error"):
        res["red"].append("門が転んだ：%s" % str(gate["error"])[:300])
    if gate.get("png"):
        res["png"] = gate["png"]

    # ★画面の言葉を信じない。台帳が増えたかで判定する
    time.sleep(3)
    after = ledger_ids()
    fueta = [i for i in after - before if i]
    res["fuetaCount"] = len(fueta)
    res["fuetaIds"] = fueta
    res["diag"].append("投げた後の台帳 %d行（増えた %d件）" % (len(after), len(fueta)))
    gamen_ok = sum(1 for x in res["gamen"] if x.get("ok"))
    res["gamenOk"] = gamen_ok
    # ④パターン：URLの分が3本＋ひとことだけの分は nagekomi_shiji（別の置き場）なので台帳は増えない
    res["kitai"] = {"台帳に増える本数": 3, "指示として残る本数": 1}
    if len(fueta) < 3:
        res["red"].append("★台帳が %d件しか増えていない（3件のはず）＝スマホの経路が通っていない"
                          % len(fueta))
    if gamen_ok < 4:
        res["red"].append("★画面が「届いた」と出なかったパターンがある（%d/4）" % gamen_ok)
    res["ok"] = not res["red"]
    return res


def run_job(payload):
    p = payload or {}
    out = run(box=p.get("box"))
    out.setdefault("totalYen", 0.0)
    out.setdefault("ok", not out.get("red"))
    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--box")
    a = ap.parse_args()
    print(json.dumps(run(box=a.box), ensure_ascii=False, indent=1)[:6000])
    return 0


if __name__ == "__main__":
    sys.exit(main())
